### PR TITLE
Feature/bookingcancel

### PR DESCRIPTION
**Pull request by purushothamanTR**
booking cancel feature added webapi and mvc

### CODE REVIEW

File: BookingSystem.BAL/Service/BookingPartialClass/UpdateBooking.cs [hunk 1]
```diff
@@ -14,5 +14,15 @@ public async Task<BookingDataVM> UpdateBookingAsync(BookingDataVM booking)
 
             return booking;
         }
+
+        public async Task<BookingCancelVM> CancelBookingAsync(BookingCancelVM booking)
+        {
+            await _attributeRepository.CancelBookingAsync(booking).ConfigureAwait(false);
+            await _refundRepository.SaveRefundTransactionAsync(booking).ConfigureAwait(false);
+            //await _addressRepository.UpdateAddressAsync(booking.Attributes.Customers.Address).ConfigureAwait(false);
```
Comment (manic1990) [inline BookingSystem.BAL/Service/BookingPartialClass/UpdateBooking.cs:22]: Remove unwanted lines.

File: BookingSystem.DAL/Repository/AttributeRepository.cs [hunk 2]
```diff
@@ -50,5 +50,16 @@ public async Task<AttributesVM> SaveAttributeAsync(AttributesVM model)
             return model;
 
         }
+
+        public async Task<BookingCancelVM> CancelBookingAsync(BookingCancelVM model)
+        {
+
+            var data = _dataContext.Attributes.SingleOrDefault(i => i.AttributeID == model.attributeid);
```
Comment (manic1990) [inline BookingSystem.DAL/Repository/AttributeRepository.cs:57]: Use FirstOrDefault();

File: BookingSystem/APIClient/Repository/APIClientBookingPartialClass/UpdateBooking.cs [hunk 3]
```diff
@@ -30,5 +30,24 @@ public async Task<BookingDataVM> UpdateBookingAsync(BookingDataVM data)
...
+            var uri = new Uri(httpClient.BaseAddress + Constant.Cancel_Booking_Async);
+            JsonContent content = JsonContent.Create(data);
+            using (var response = await httpClient.PostAsync(uri, content))
+            {
+                if (response.StatusCode == HttpStatusCode.OK)
+                {
+                    var responses = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                    var responseData = JsonConvert.DeserializeObject<object>(responses);
+                    var bookingData = JsonConvert.DeserializeObject<ResponseModel>(responseData.ToString());
+                    var result = JsonConvert.DeserializeObject<BookingCancelVM>(bookingData.response.ToString());
```
Comment (manic1990) [inline BookingSystem/APIClient/Repository/APIClientBookingPartialClass/UpdateBooking.cs:45]: Check null for bookingData.response

File: BookingSystem/Controllers/BookingController.cs [hunk 4]
```diff
@@ -276,5 +276,46 @@ public async Task<IActionResult> Get()
...
+                    _logger.LogInformation(nameof(BookingController) + "." + nameof(Booking) + complete);
+
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogError(ex.Message);
+                }
+            }
+            return RedirectToAction(nameof(Get));
+            //return View(viewBookingData);
```
Comment (manic1990) [inline BookingSystem/Controllers/BookingController.cs:300]: remove unwanted line

File: BookingSystem.BAL/Service/BookingPartialClass/UpdateBooking.cs (hunk 1 above)
Comment (purushothamanTR) [reply BookingSystem.BAL/Service/BookingPartialClass/UpdateBooking.cs]: removed unwanted lines

File: BookingSystem.DAL/Repository/AttributeRepository.cs (hunk 2 above)
Comment (purushothamanTR) [reply BookingSystem.DAL/Repository/AttributeRepository.cs]: use FirstOrDefault()

File: BookingSystem/APIClient/Repository/APIClientBookingPartialClass/UpdateBooking.cs (hunk 3 above)
Comment (purushothamanTR) [reply BookingSystem/APIClient/Repository/APIClientBookingPartialClass/UpdateBooking.cs]: Check null for bookingData.response

File: BookingSystemAPI/Controllers/BookingController.cs [hunk 5]
```diff
@@ -200,5 +200,33 @@ public async Task<ActionResult> GetBookingByIdAsync(long bookingId)
...
+        [HttpPost]
+        [Route("cancelBooking")]
+        public async Task<ActionResult> CancelBookingAsync(BookingCancelVM booking)
+        {
+            _logger.LogInformation(nameof(BookingController) + nameof(CancelBookingAsync), started);
+            var response = new ResponseModel();
+            try
+            {
+                _logger.LogInformation(nameof(BookingService) + cancelBooking, started);
+                var result = await _bookingService.CancelBookingAsync(booking);
```
[the remark this answers is not in the document]
Comment (purushothamanTR) [reply BookingSystemAPI/Controllers/BookingController.cs]: Use .ConfigureAwait(false)

File: BookingSystem.Tests/APIController.Tests/APIBookingController.Tests/CancelBookingTest.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+            var mockLoggerFactory = new Mock<ILoggerFactory>();
+            mockLoggerFactory.Setup(x => x.CreateLogger(It.IsAny<string>())).Returns(() => _mockLog.Object);
+
+            _mockBookingService.Setup(repo => repo.CancelBookingAsync(resuestData)).ReturnsAsync(data);
+            var _bookingController = new BookingController(_mockBookingService.Object, mockLoggerFactory.Object);
+
+            // Act
+            var result = _bookingController.CancelBookingAsync(resuestData);
+
```
Comment (Backiaraj07) [inline BookingSystem.Tests/APIController.Tests/APIBookingController.Tests/CancelBookingTest.cs:55]: Use await

File: BookingSystem.Tests/BAL.Tests/BookingServiceTest/SaveBookingTest.cs [hunk 7]
```diff
@@ -63,7 +65,7 @@ public void SaveBooking_ReturnOKResult_WithBookingId()
                                                    _roomRepository.Object, _venueRepository.Object, _cardRepository.Object,
                                                    _paymentCardRepository.Object, _realtionshipRepository.Object, _addressRepository.Object,
                                                    _rateCardRepository.Object, _chargesRepository.Object,
-                                                   _paymentAttemptsRepository.Object, _paymentReversalRepository.Object, _guestProfileRepository.Object, _hotelRepository.Object);
+                                                   _paymentAttemptsRepository.Object, _paymentReversalRepository.Object, _guestProfileRepository.Object, _hotelRepository.Object, _refundRepository.Object);
 
             // Act
             var result = _bookingService.SaveBookingAsync(data);
```
Comment (Backiaraj07) [inline BookingSystem.Tests/BAL.Tests/BookingServiceTest/SaveBookingTest.cs:71]: Use Await

File: BookingSystem/Controllers/BookingController.cs [hunk 8]
```diff
@@ -276,5 +276,29 @@ public async Task<IActionResult> Get()
             return View();
         }
 
+
+        [HttpGet]
+        public async Task<IActionResult> CancelBooking(BookingCancelVM model
+            )
+        {
```
Comment (Backiaraj07) [inline BookingSystem/Controllers/BookingController.cs:283]: Make it proper alignment

**Review by Backiaraj07 — COMMENTED** [review]
Comments added need to be resolved

**Review by manic1990 — APPROVED** [review]

File: BookingSystem.Tests/APIController.Tests/APIBookingController.Tests/CancelBookingTest.cs (hunk 6 above)
Comment (purushothamanTR) [reply BookingSystem.Tests/APIController.Tests/APIBookingController.Tests/CancelBookingTest.cs:55]: await added

File: BookingSystem.Tests/BAL.Tests/BookingServiceTest/SaveBookingTest.cs (hunk 7 above)
Comment (purushothamanTR) [reply BookingSystem.Tests/BAL.Tests/BookingServiceTest/SaveBookingTest.cs:71]: await added

File: BookingSystem/Controllers/BookingController.cs (hunk 8 above)
Comment (purushothamanTR) [reply BookingSystem/Controllers/BookingController.cs:283]: alignment changed

File: BookingSystem/Controllers/BookingController.cs [hunk 9]
```diff
@@ -276,5 +276,28 @@ public async Task<IActionResult> Get()
             return View();
         }
 
+
+        [HttpGet]
+        public async Task<IActionResult> CancelBooking(BookingCancelVM model)
+        {
+            _logger.LogInformation(nameof(BookingController) + "." + nameof(Booking) + start);
+            var viewBookingData = new ViewBookingData();
```
Comment (Backiaraj07) [inline BookingSystem/Controllers/BookingController.cs:284]: Dont keep hardcode values in controller, keep it in constant class

image

**Review by Backiaraj07 — COMMENTED** [review]
Comments added

File: BookingSystem/Controllers/BookingController.cs [hunk 10]
```diff
@@ -276,5 +276,28 @@ public async Task<IActionResult> Get()
             return View();
         }
 
+
+        [HttpGet]
+        public async Task<IActionResult> CancelBooking(BookingCancelVM model)
+        {
+            _logger.LogInformation(nameof(BookingController) + "." + nameof(Booking) + start);
```
Comment (bhairab15) [inline BookingSystem/Controllers/BookingController.cs:283]: Can we use string interpolation here?

File: BookingSystem/Controllers/BookingController.cs [hunk 11]
```diff
@@ -276,5 +276,28 @@ public async Task<IActionResult> Get()
...
+            var viewBookingData = new ViewBookingData();
+            if (model.bookingId > 0)
+            {
+                try
+                {
+                    await _clientRepository.CancelBookingAsync(model).ConfigureAwait(false);
+                    _logger.LogInformation(nameof(BookingController) + "." + nameof(Booking) + complete);
```
Comment (bhairab15) [inline BookingSystem/Controllers/BookingController.cs:290]: please use string interpolation

File: BookingSystem/Controllers/BookingController.cs (hunk 9 above)
Comment (purushothamanTR) [reply BookingSystem/Controllers/BookingController.cs]: fixed

File: BookingSystem/Controllers/BookingController.cs (hunk 10 above)
Comment (purushothamanTR) [reply BookingSystem/Controllers/BookingController.cs]: changed as string interpolation

File: BookingSystem/Controllers/BookingController.cs (hunk 11 above)
Comment (purushothamanTR) [reply BookingSystem/Controllers/BookingController.cs]: changed as string interpolation

File: BookingSystem.Tests/BAL.Tests/BookingServiceTest/CancelBookingTest.cs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            _paymentReversalRepository = new Mock<IPaymentReversalRepository>();
+            _guestProfileRepository = new Mock<IGuestProfileRepository>();
+            _refundRepository = new Mock<IRefundRepository>();
+            _hotelRepository = new Mock<IHotelRepository>();
+        }
+
+        /// <summary>
+        /// Test method for CancelBookingAsync
+        /// </summary>
+        /// <returns>data</returns>
```
Comment (bhairab15) [inline BookingSystem.Tests/BAL.Tests/BookingServiceTest/CancelBookingTest.cs:81]: please remove comments

File: BookingSystem.Tests/Integration.Tests/BookingRepositoryTest/CancelBookingTest.cs [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        public static IConfiguration InitConfiguration()
+        {
+            var config = new ConfigurationBuilder()
+                .AddJsonFile("appunittestsettings.json")
+                .Build();
+            return config;
+        }
+
+        /// <summary>
+        /// Test method for UpdateBookingAsync
```
Comment (bhairab15) [inline BookingSystem.Tests/Integration.Tests/BookingRepositoryTest/CancelBookingTest.cs:64]: Remove comments

**Review by Backiaraj07 — APPROVED** [review]
Code reviewed

File: BookingSystem.Tests/Integration.Tests/BookingRepositoryTest/CancelBookingTest.cs (hunk 13 above)
Comment (purushothamanTR) [reply BookingSystem.Tests/Integration.Tests/BookingRepositoryTest/CancelBookingTest.cs]: removed

File: BookingSystem.Tests/BAL.Tests/BookingServiceTest/CancelBookingTest.cs (hunk 12 above)
Comment (purushothamanTR) [reply BookingSystem.Tests/BAL.Tests/BookingServiceTest/CancelBookingTest.cs]: Removed comments

File: BookingSystem/Controllers/BookingController.cs [hunk 14]
```diff
@@ -170,7 +172,7 @@ public async Task<IActionResult> GetHotelsByVenueId(int venueId)
             }
             catch (Exception ex)
             {
-                _logger.LogError(ex.Message);
+                _logger.LogError($"{ex.Message}");
```
Comment (bhairab15) [inline BookingSystem/Controllers/BookingController.cs:175]: If there is no concatenation we do not need interpolation

**Review by bhairab15 — APPROVED** [review]

**Pull request merged**